coding/javascript.md

```diff
@@ -37,6 +37,28 @@ We use [Yarn](https://yarnpkg.com/en) for dependency management.
 
 We use the [Jest](https://jestjs.io) unittest framework.
 
+#### General Testing Advice
+* Avoid using `beforeEach`. It should be clear from reading a test in isolation what state (e.g. fixtures, mocks) it is in. Using `beforeEach` creates a number of issues:
```

Comment:
Not sure if I agree with this.

Yes, ideally a test function by itself should be self explanatory and could contain both the setup/context of the test and the test itself.
But if there are multiple tests to be run on the same or very similar setup it's worth (and should be encouraged) to remove the duplication and use `beforeEach`.
Sometimes setup may be quite big and complicated (preparing DOM structure, mocking requests, etc). Having this whole code in the test function makes the 'test' part of it harder to read. In build.snapcraft backend tests where we use nock for mocking requests test setup is sometimes much longer then tests themselves (mocking multiple requests to various endpoints and services) - it would be very error prone to have it duplicated in every test just to avoid using `beforeEach`.

I agree that these are valid examples of some issues usage of `beforeEach` may cause and developers writing tests should be aware of them, but having "avoid using `beforeEach`" as very first testing advice seems to strong to me. `beforeEach` is useful and I think it should be used to avoid duplication and to separate test logic from test setup (especially if setup is shared by multiple tests run in the same context).

Reply:
@bartaz appreciate the response. How about a compromise where we suggest that tests sharing the *same* setup are within the same `describe` scope, but as soon as the fixtures diverge they should have a new `describe` scope with their own `beforeEach` context?

Reply:
> But these are all very specific implementation details of tests. I'm not sure how much we need to codify it here as a practice.

I just think it is a generally a problematic code smell, and have worked in code bases where it has lead to very confusing tests, particularly when coupled with testcase inheritance.

Reply:
OK. I guess in our current JS codebase we don't have enough tests and experience in writing tests for our JS code, so first of all we need to start to write more tests and build some practices around that.

> Use `beforeEach` sparingly and with care.

This sounds better than "avoid" (at least more permissive). Maybe later we could find or write more detailed guidelines for writing tests (what to group in `describe`, when it is ok to use `beforeEach` and when to keep setup in test itself, etc) and link from here.

Reply:
Sorry this branch has been sitting around and getting rather musty. I've updated the advice around `beforeEach` now to be far less dogmatic. Did you want to have another look @bartaz? Thanks!